zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -565,18 +600,19 @@ impl UpdateWith<ContextuallyValidBlock> for Chain {
             };
 
             // add key `transaction.hash` and value `(height, tx_index)` to `tx_by_hash`
+            let transaction_location = TransactionLocation::from_usize(height, transaction_index);
             let prior_pair = self
                 .tx_by_hash
-                .insert(transaction_hash, (height, transaction_index));
-            assert!(
-                prior_pair.is_none(),
+                .insert(transaction_hash, transaction_location);
+            assert_eq!(
+                prior_pair, None,
                 "transactions must be unique within a single chain"
             );
 
-            // add the utxos this produced
-            self.update_chain_tip_with(new_outputs)?;
```

Comment:
This was one bug: `new_outputs` contains all the UTXOs for the block, but it was indexed for every transaction.